Use allow-list for mobile editor settings endpoint

mkevins opened the pull request:
### Description

This PR introduces an "allow-list" for the mobile endpoint for block editor settings. This gives us a greater degree of control over what the client apps are consuming, and helps avoid prematurely introducing changes which do not yet have support in the mobile apps.

## What?
<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

mkevins committed:
Use allow-list for mobile editor settings endpoint

dcalhoun committed:
refactor: Format php

Address linter errors and warnings.

dcalhoun committed:
refactor: Guard mobile allow list utility against redeclaration

Address linter error.

dcalhoun commented on lib/experimental/block-editor-settings-mobile.php:
```diff
@@ -33,7 +67,64 @@ function gutenberg_get_block_editor_settings_mobile( $settings ) {
 	}
 
-	return $settings;
+	return keep_supported_block_editor_settings_mobile(
```
I think `filter_by_supported_block_editor_settings_mobile` better communicates the intent. WDYT? Likely merely a personal preference, feel free to disregard.

mkevins replied:
I like it, actually. Initially, I had "filter" in the name, and then second-guessed myself since the semantics of `addFilter` can mean more than selectively removing items, etc. (more akin to a middleware), but I was probably overthinking it 🤷‍♂️, and filter is just fine in the name. 👍

mkevins replied:
Resolved w/ https://github.com/WordPress/gutenberg/pull/56424/commits/8f072e3fd0acbc912331843050ea1fdefef36cbb

dcalhoun commented on lib/experimental/block-editor-settings-mobile.php:
```diff
@@ -33,7 +67,64 @@ function gutenberg_get_block_editor_settings_mobile( $settings ) {
+			'disableCustomGradients'           => true,
+			'disableLayoutStyles'              => true,
+			'enableCustomLineHeight'           => true,
+			'enableCustomSpacing'              => true,
+			'enableCustomUnits'                => true,
+			'colors'                           => true,
+			'fontSizes'                        => true,
+			'__experimentalFeatures'           => array(
+				'appearanceTools'               => true,
+				'useRootPaddingAwareAlignments' => true,
+				'border'                        => true,
+				'color'                         => true,
+				'shadow'                        => true,
+				'spacing'                       => true,
+				'typography'                    => array(
+					'dropCap'        => true,
+					'fontSizes'      => true,
+					'fontStyle'      => true,
+					'fontWeight'     => true,
+					'letterSpacing'  => true,
+					'textColumns'    => true,
+					'textDecoration' => true,
+					'textTransform'  => true,
+					'writingMode'    => true,
+				),
+				'blocks'                        => true,
+				'background'                    => true,
+				'dimensions'                    => true,
+				'position'                      => true,
+			),
+			'gradients'                        => true,
+			'disableCustomSpacingSizes'        => true,
+			'spacingSizes'                     => true,
+			'__unstableIsBlockBasedTheme'      => true,
+			'localAutosaveInterval'            => true,
+			'__experimentalDiscussionSettings' => true,
+			'__experimentalDashboardLink'      => true,
+			'__experimentalEnableQuoteBlockV2' => true,
+			'__experimentalEnableListBlockV2'  => true,
+		)
```
I'll preface this thought with the fact that it is has been a very long time since I last wrote PHP. I'm less familiar with its current syntax or paradigms. 

Would it make sense to hoist this array value and assign it to a variable/constant? I wonder if it would make the code a bit easier to comprehend? WDYT?

dcalhoun replied:
After reviewing the response from the API via `curl`, the values in this array appears to match the network response. I have not tested functionality in the editor, but this list makes sense to me.

mkevins replied:
> it is has been a very long time since I last wrote PHP. I'm less familiar with its current syntax or paradigms.

I'm in the same boat 😅 

> Would it make sense to hoist this array value and assign it to a variable/constant?

That does make sense to me. 👍

mkevins replied:
Resolved w/ https://github.com/WordPress/gutenberg/pull/56424/commits/7f3d7052ebc2bea9847708041bd9a4d6118cb00d

dcalhoun reviewed (commented):
I took the liberty of refactoring and formatting the code to address the lint errors and warning, as the number of inline errors made review a bit difficult. 

I have not yet tested the code, but at a cursory glance the changes seems sound to me.

dcalhoun committed:
fix: Avoid error from selecting non-existent settings key

If the settings do not include an item listed in the allow list, we must
avoid an attempt to reference the non-existent settings key.

dcalhoun committed:
fix: Add `__experimentalStyles` key to the allow list

This appears to be expected by the existing tests. However, it is
unclear if all or a subset of the nested keys are expected or required.

dcalhoun commented on lib/experimental/block-editor-settings-mobile.php:
```diff
@@ -33,7 +71,71 @@ function gutenberg_get_block_editor_settings_mobile( $settings ) {
+			'enableCustomUnits'                => true,
+			'colors'                           => true,
+			'fontSizes'                        => true,
+			'__experimentalStyles'             => array(
+				'elements'   => true,
+				'spacing'    => true,
+				'blocks'     => true,
+				'color'      => true,
+				'typography' => true,
+			),
```
At the moment, I am not certain what, if any, of `__experimentalStyles` is utilized. However, it appears to be expected in the current test.

https://github.com/WordPress/gutenberg/blob/e3a6917ff0353bba5d40660acb1bc273de2b2541/phpunit/experimental/block-editor-settings-mobile-test.php#L64

mkevins replied:
I'm not entirely sure where this is used, either, but I recall that this is specifically handled (as `rawStyles`) in the app's db, so good idea to add it. Fwiw, I generated the allow-list from an existing response from a specific site, so there may be other things that are missing in this list that would (and possibly should) appear for other site-types.

dcalhoun commented on lib/experimental/block-editor-settings-mobile.php:
```diff
@@ -33,7 +71,71 @@ function gutenberg_get_block_editor_settings_mobile( $settings ) {
+			'__unstableIsBlockBasedTheme'      => true,
+			'localAutosaveInterval'            => true,
+			'__experimentalDiscussionSettings' => true,
+			'__experimentalDashboardLink'      => true,
```
I didn't find many references to `__experimentalDashboardLink`. It appears to be a hook to customize the back button in the full site editor, which I do not believe is relevant to the the mobile editor. My thought is that this could be excluded. WDYT?

mkevins replied:
Resolved w/ https://github.com/WordPress/gutenberg/pull/56424/commits/9c19187a4394697fa3e676b44462a10101063e66

dcalhoun commented on lib/experimental/block-editor-settings-mobile.php:
```diff
@@ -33,7 +71,71 @@ function gutenberg_get_block_editor_settings_mobile( $settings ) {
-	return $settings;
+	return keep_supported_block_editor_settings_mobile(
+		$settings,
+		array(
```
After reviewing the response from the API via `curl`, the values in this array appear to match the network response. I have not tested functionality in the editor, but this list makes sense to me.

mkevins commented on lib/experimental/block-editor-settings-mobile.php:
```diff
@@ -23,6 +23,10 @@ function keep_supported_block_editor_settings_mobile( $initial_array, $allow_lis
 		$result = array();
 
 		foreach ( $allow_list_array as $key => $value ) {
+			if ( ! array_key_exists( $key, $initial_array ) ) {
+				continue;
+			}
+
```
Thank you for adding this check!  Originally, I had the foreach traversing through the response instead of the allow-list, and when I switched it to the current approach this was in the back of my mind. I just forgot to add this logic before drafting the PR 🤦‍♂️.

mkevins committed:
Rename mobile settings allow-list filter function

mkevins committed:
Extract mobile editor settings allow list to a constant

mkevins committed:
Remove experimental dashboard link from the mobile allow list

mkevins commented:
Thank you @dcalhoun  for your review (and for your helpful commits)! I've pushed a few more commits to address your feedback, so this is ready for another pass.

Note: I have not yet marked this as "ready for review" because I haven't had time to add some unit tests. But, if you have further feedback about the approach, it is welcome 🙂 .

mkevins committed:
Extract mobile editor settings filter to a separate file

The important part of this is that the filter is a separate filter.
Incidentally, this also fixes a bug in which the previous draft
implementation returned a filtered response even for non-mobile
contexts.

mkevins committed:
Load mobile-specific filter for the editor settings endpoint

dcalhoun committed:
refactor: Format php

Appease linter errors via `npm run format:php`.

dcalhoun commented:
> Thank you @dcalhoun for your review (and for your helpful commits)! I've pushed a few more commits to address your feedback, so this is ready for another pass.
> 
> Note: I have not yet marked this as "ready for review" because I haven't had time to add some unit tests. But, if you have further feedback about the approach, it is welcome 🙂 .

@mkevins the implementation looks sound to me. I do not have further feedback at this time. 

I pushed a commit addressing the PHP unit test failures by running `npm run format:php`. The E2E test failures appear unrelated, we should likely rebase this branch atop the latest `trunk` to see if it resolves the failures. 

From what I gathered, a version of these changes were deployed internally. So, for better or worse, I suppose further stability testing of these changes is happening live. If that appears to go well, then it increases confidence that these proposed changes include all the necessary attributes for the allow list.

geriux commented on lib/experimental/block-editor-settings-mobile-allowed.php:
```diff
@@ -0,0 +1,138 @@
+				'color'      => true,
+				'typography' => true,
+			),
+			'__experimentalFeatures'           => array(
```
Let's also add the `custom` attribute which is used for [custom variables or other settings](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#css-custom-properties-presets-custom). The app uses it to get values for custom variables.
